contractsV5/FeeHandler.sol

```diff
@@ -0,0 +1,189 @@
+        );
+        _;
+    }
+
+
+    function encodeData(uint _burn, uint _reward, uint _epoch, uint _expiryBlock) public pure returns (uint) {
+        return (((((_burn << BITS_PER_PARAM) + _reward) << BITS_PER_PARAM) + _epoch) << BITS_PER_PARAM) + _expiryBlock;
+    }
+
+    function decodeData() public view returns(uint, uint, uint, uint) {
+        uint expiryBlock = brrAndEpochData & (1 << BITS_PER_PARAM) - 1;
+        uint epoch = (brrAndEpochData / (1 << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        uint rewardInBPS = (brrAndEpochData / (1 << BITS_PER_PARAM << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        uint burnInBPS = (brrAndEpochData / (1 << BITS_PER_PARAM << BITS_PER_PARAM << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        return (burnInBPS, rewardInBPS, epoch, expiryBlock);
+    }
+
+    event AccumulateReserveRebate(address rebateWallet, uint rebateAmountWei);
+
+    // Todo: future optimisation to accumulate rebates for 2 rebate wallet
+    // encode totals, 128 bits per reward / rebate
+    function handleFees(address[] calldata eligibleWallets, uint[] calldata rebatePercentages) external payable onlyKyberNetwork returns(bool) {
+        // Decoding BRR data
+        (uint burnInBPS, uint rewardInBPS, uint epoch, uint expiryBlock) = decodeData();
+        uint rebateInBPS = BPS - rewardInBPS - burnInBPS;
+
+        // Check current block number
+        if(block.number > expiryBlock) {
+            (burnInBPS, rewardInBPS, rebateInBPS, epoch, expiryBlock) = kyberDAOContract.getLatestBRRData();
+
+            // Update brrAndEpochData
+            brrAndEpochData = encodeData(burnInBPS, rewardInBPS, epoch, expiryBlock);
+        }
+
+        uint rebateWei = rebateInBPS * msg.value / BPS;
+        uint rewardWei = rewardInBPS * msg.value / BPS;
+        for(uint i = 0; i < eligibleWallets.length; i ++) {
+            // Internal accounting for rebates per reserve wallet (totalRebatesPerRebateWallet)
+            totalRebatesPerRebateWallet[eligibleWallets[i]] = rebateWei * rebatePercentages[i] / 100;
+            // Internal accounting for total rebates (totalRebates)
+            totalRebates += rebateWei * rebatePercentages[i] / 100;
+            // Internal accounting for reward per epoch (totalRewardsPerEpoch)
+            totalRewardsPerEpoch[epoch] += rewardWei;
+            // Internal accounting for total rewards (totalRewards)
+            totalRewards += rewardWei;
+
+            emit AccumulateReserveRebate(eligibleWallets[i], rebateWei * rebatePercentages[i] / 100);
+        }
+        return true;
+    }
+
+    event DistributeRewards(address staker, uint amountWei);
+
+    function claimStakerReward(address staker, uint percentageInPrecision, uint epoch) public onlyDAO returns(uint) {
+        // Amount of reward to be sent to staker
+        uint amount = totalRewardsPerEpoch[epoch] * percentageInPrecision / 10**18;
```

Comment:
in utils we should have PRECISION constant.

Reply:
updated and will push fix when other comments are fixed